File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-dotnet/05_AzureServiceBus.cs
// TRIPLE:
//   Skill: azure-servicebus-dotnet
//   Prompt: "Using azure-servicebus-dotnet skill, write C# code that sends a message to
//            a Service Bus queue and receives with peek-lock + complete/dead-letter."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-dotnet/skills/azure-servicebus-dotnet

using System.Text.Json;
using Azure.Identity;
using Azure.Messaging.ServiceBus;

var ns = Environment.GetEnvironmentVariable("SERVICEBUS_NAMESPACE")!;
var queue = "document-ingestion";
await using var client = new ServiceBusClient(ns, new DefaultAzureCredential());

// Send
await using var sender = client.CreateSender(queue);
var body = JsonSerializer.Serialize(new { blob_url = "https://storage/doc.pdf", action = "index" });
await sender.SendMessageAsync(new ServiceBusMessage(body) { ContentType = "application/json" });
Console.WriteLine($"Sent to {queue}");

// Receive with peek-lock
await using var receiver = client.CreateReceiver(queue);
var msg = await receiver.ReceiveMessageAsync(TimeSpan.FromSeconds(10));
if (msg != null)
{
    try
    {
        var doc = JsonSerializer.Deserialize<JsonElement>(msg.Body.ToString());
        Console.WriteLine($"Received: {doc}");
        await receiver.CompleteMessageAsync(msg);
        Console.WriteLine("  → completed");
    }
    catch (JsonException)
    {
        await receiver.DeadLetterMessageAsync(msg, "parse_failure", "Invalid JSON");
        Console.WriteLine("  → dead-lettered");
    }
}

File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-java/05_AzureServiceBus.java
// TRIPLE:
//   Skill: azure-servicebus-java
//   Prompt: "Using azure-servicebus-java skill, write Java code that sends and receives messages with peek-lock."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-java/skills/azure-servicebus-java

import com.azure.identity.DefaultAzureCredentialBuilder;
import com.azure.messaging.servicebus.*;

import java.time.Duration;

public class AzureServiceBus {

    public static void main(String[] args) {
        String namespace = System.getenv("AZURE_SERVICEBUS_NAMESPACE");
        String queueName = System.getenv("AZURE_SERVICEBUS_QUEUE") != null
                ? System.getenv("AZURE_SERVICEBUS_QUEUE") : "skill-demo-queue";

        // Send a message
        ServiceBusSenderClient sender = new ServiceBusClientBuilder()
                .fullyQualifiedNamespace(namespace)
                .credential(new DefaultAzureCredentialBuilder().build())
                .sender()
                .queueName(queueName)
                .buildClient();

        ServiceBusMessage message = new ServiceBusMessage("{\"action\":\"process\"}")
                .setContentType("application/json")
                .setSubject("skill-demo");
        sender.sendMessage(message);
        System.out.println("Message sent.");
        sender.close();

        // Receive with peek-lock (default)
        ServiceBusReceiverClient receiver = new ServiceBusClientBuilder()
                .fullyQualifiedNamespace(namespace)
                .credential(new DefaultAzureCredentialBuilder().build())
                .receiver()
                .queueName(queueName)
                .buildClient();

        for (ServiceBusReceivedMessage msg : receiver.receiveMessages(1, Duration.ofSeconds(5))) {
            System.out.printf("Received: %s%n", msg.getBody().toString());
            try {
                // Process then complete
                receiver.complete(msg);
                System.out.println("Message completed.");
            } catch (Exception e) {
                receiver.deadLetter(msg, new DeadLetterOptions()
                        .setDeadLetterReason("ProcessingFailed")
                        .setDeadLetterErrorDescription("Demo: simulated failure"));
                System.out.println("Message dead-lettered.");
            }
        }

        receiver.close();
    }
}
